src/vehicle_autodrive.cpp

```diff
@@ -1,25 +1,1202 @@
+ * to a large obstacle that can't be driven around from that position (a dead end) and instead
+ * we always choose to cross OMT boundaries in a position from which we are guaranteed to be
+ * able to continue onward.
+ *
+ * The pathfinding algorithm being employed is A*. The graph being searched is made up of nodes
+ * denoted by x and y coordinates as well as a direction of travel. Since steering changes the
+ * vehicle's direction in increments of 15 degrees there are 360/15 possible orientations. By
+ * limiting navigation to 2 OMTs at a time we have a maximum search space of 2*24*24*24 nodes.
+ * In practice, A* finds a good path after visiting fewer than 50 nodes due to the heuristic
+ * function that estimates which nodes are most likely to lead to the destination (which is
+ * based on proximity to the "exit zone" leading to the next OMT and whether the vehicle is
+ * pointing towards the middle of that zone). Although there can be many paths that we could
+ * take to get through to the next OMT, we just take the first one we find since it tends to
+ * be the most direct and is usually near-optimal. When the way is blocked we'll either give
+ * up after exhausting the search space or after we hit a limit on the search length.
+ *
+ * In order to build the search graph for A*, we consider every possible map tile and orientation
+ * in the search zone and we figure out whether the vehicle's pivot point could be placed there
+ * without the vehicle overlapping some obstacle (a collision-free placement). The pivot point
+ * is the part of the vehicle that moves the least while steering and is usually between the
+ * wheels of the vehicle, closer to the non-steerable ones. The pathfinding problem reduces to
+ * finding a path for the pivot point to follow through the set of possible positions and
+ * orientations (nodes). In order to figure out the possible connections between nodes (the
+ * edges of the graph) we consider a limited set of actions that the driver can perform at
+ * each location: nothing, steer left or steer right (possibly more than once). Given a current
+ * node, current speed and amount of steering we can predict where the vehicle's pivot point
+ * will end up in 1 turn and connect the 2 nodes.
+ *
+ * In order to keep the search space small we only consider paths traveled at constant speed
+ * (otherwise we'd have to add speed as an extra attribute in every node). When we need to
+ * compute a new path we first try to use the maximum safe autodrive speed (which is typically
+ * much less than the vehicle's max speed). If we can't find a path we cut the speed in half and
+ * try again, repeating until we succeed or reach the minimum speed. When the target speed is
+ * different from the current speed of the vehicle we try to estimate how long it will take to
+ * reach that speed and take that into account when simulating the first few edges (they'll be
+ * shorter or longer than the rest).
```

Comment:
What are the limits on the accuracy of this? (If improvements can be made in the accuracy, this may be a way to differentiate drivers of different skill.)

Reply:
The loss of accuracy results from 2 sources:
1) expressing speeds in terms of tiles/s.
2) ignoring "remainder" moves, or fractional tile displacement

1) The vehicle move code uses hundredth-of-a-mile/s for speed, with 1tile/s=400hmps. When vehicle motion is simulated by autodrive it uses tiles/s in order to simplify the calculations. This is fine when the vehicle reaches the desired speed, but while it's accelerating it will have speeds that cannot be accurately expressed in tps units.

2) The vehicle move code displaces each vehicle a whole number of tiles, but if the speed was not an exact multiple of 400hmps it also saves a "remainder" (like a fraction of a tile), which will be added next turn. So a vehicle travelling at 600hmps (1.5 tps) will move 1 tile in one turn and 2 tiles the next, and so on. In autodrive vehicles always move a constant number of tiles per turn for a given speed.

These accuracy losses can be reduced or eliminated by adding more variables to the simulation, but I don't want to complicate this PR further.